### PR TITLE
refactor and fix edge case in gdal_source

### PR DESCRIPTION
**Pull request by jdroenner**
refactored tile loading to make it more understandable

fixed edge case when request is inside a single pixel of the original dataset

removed partition from GdalDatasetParameters. Added grid_shape instead.

### CODE REVIEW

**Comment by jdroenner** [comment]
[gdal_source edge case wenn request in einem pixel liegt](https://trello.com/c/9gJ4WeWK/263-gdalsource-edge-case-wenn-request-in-einem-pixel-liegt)

File: datatypes/src/raster/geo_transform.rs [hunk 1]
```diff
@@ -111,15 +113,30 @@ impl GeoTransform {
...
-        let end: GridIdx2D = [max(ul[0], lr[0]), max(ul[1], lr[1])].into();
+    pub fn spatial_to_grid_bounds(
+        &self,
+        spatial_partition: &SpatialPartition2D,
+    ) -> GridBoundingBox2D {
+        //let snapped = spatial_partition.snap_to_grid(self.origin_coordinate, self.spatial_resolution());
+        let GridIdx([ul_y, ul_x]) = self.coordinate_to_grid_idx_2d(spatial_partition.upper_left());
+        let GridIdx([lr_y, lr_x]) = self.coordinate_to_grid_idx_2d(spatial_partition.lower_right()); // this is the next pixel!
+
+        // dbg!(&spatial_partition, &snapped);
```
Comment (michaelmattig) [inline datatypes/src/raster/geo_transform.rs:126]: delete

File: operators/src/source/gdal_source.rs [hunk 2]
```diff
@@ -1142,6 +1148,31 @@ mod tests {
...
+        let output_shape: GridShape2D = [8, 8].into();
+        // shift world bbox one pixel up and to the left
+        let (x_size, y_size) = (0.000_000_000_01, 0.000_000_000_01);
+        let output_bounds = SpatialPartition2D::new(
+            (-116.22222, 66.66666).into(),
+            (-116.22222 + x_size, 66.66666 - y_size).into(),
+        )
+        .unwrap();
+
+        dbg!(&output_bounds);
```
Comment (michaelmattig) [inline operators/src/source/gdal_source.rs:1162]: remove

**Review by michaelmattig — CHANGES_REQUESTED** [review]
.

File: datatypes/src/raster/geo_transform.rs [hunk 3]
```diff
@@ -111,15 +113,30 @@ impl GeoTransform {
...
+
+        debug_assert!(ul_x <= lr_x);
+        debug_assert!(ul_y <= lr_y);
+
+        let lr_x_inc = max(ul_x, lr_x - 1);
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/geo_transform.rs:131]: inc for inclusive

File: operators/src/source/gdal_source.rs [hunk 4]
```diff
@@ -140,12 +141,26 @@ pub struct GdalDatasetParameters {
     pub file_path: PathBuf,
     pub rasterband_channel: usize,
     pub geo_transform: GeoTransform,
-    pub partition: SpatialPartition2D, // the spatial partition of the dataset containing the raster data
+    pub grid_shape: GridShape2D,
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs:144]: i changed it to width height

File: operators/src/source/gdal_source.rs [hunk 5]
```diff
@@ -140,12 +141,26 @@ pub struct GdalDatasetParameters {
...
 }
 
+impl SpatialPartitioned for GdalDatasetParameters {
+    fn spatial_partition(&self) -> SpatialPartition2D {
+        let lower_right_coordinate = self.geo_transform.origin_coordinate
+            + Coordinate2D::from((
+                self.geo_transform.x_pixel_size * self.grid_shape.axis_size_x() as f64,
+                self.geo_transform.y_pixel_size * self.grid_shape.axis_size_y() as f64,
+            ));
+        SpatialPartition2D::new_unchecked(
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs:157]: changed the deserializer method to test that

File: datatypes/src/raster/geo_transform.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/geo_transform.rs]: a14864a

File: datatypes/src/raster/geo_transform.rs [hunk 6]
```diff
@@ -111,15 +113,30 @@ impl GeoTransform {
...
+        let lr_y_inc = max(ul_y, lr_y - 1);
+
+        debug_assert!(ul_x <= lr_x_inc);
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/geo_transform.rs]: a14864a

File: operators/src/source/gdal_source.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: 02f15b8

File: operators/src/source/gdal_source.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs:159]: 06eab1c

File: operators/src/source/gdal_source.rs (hunk 2 above)
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: remove dbgs

File: operators/src/source/gdal_source.rs [hunk 7]
```diff
@@ -404,64 +419,58 @@ where
...
+                &dataset_grid_bounds,
+                output_shape,
+                no_data_value,
+            )?
+            .into()
+        } else {
+            let tile_grid_bounds =
+                output_geo_transform.spatial_to_grid_bounds(&dataset_intersecton_area);
+
+            dbg!(&tile_grid_bounds);
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: 66f2ce9

File: operators/src/source/gdal_source.rs [hunk 8]
```diff
@@ -404,64 +419,58 @@ where
...
-        // let dataset_intersects_tile = dataset_bounds.intersects_bbox(&output_bounds);
-        // TODO: move to false, true case when BBOX paradox is solved
+        // check if query and dataset intersect
         let dataset_intersects_tile = dataset_bounds.intersection(&output_bounds);
-        let result_grid: GridOrEmpty2D<T> = match (dataset_contains_tile, dataset_intersects_tile) {
-            (_, None) => {
-                // TODO: refactor tile to hold an Option<GridData> and this will be empty in this case
-                if let Some(no_data) = no_data_value {
+
+        let dataset_intersecton_area = match dataset_intersects_tile {
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: 66f2ce9

File: operators/src/source/gdal_source.rs [hunk 9]
```diff
@@ -404,64 +419,58 @@ where
...
+                return Ok(GridWithProperties {
+                    grid: no_data_grid,
+                    properties,
+                });
             }
         };
 
+        let dataset_grid_bounds = geo_transform.spatial_to_grid_bounds(&dataset_intersecton_area);
+
+        let dataset_contains_tile = dataset_intersecton_area == output_bounds;
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: 66f2ce9

File: operators/src/source/gdal_source.rs [hunk 10]
```diff
@@ -404,64 +419,58 @@ where
...
+
+        dbg!(&dataset_intersecton_area, &dataset_grid_bounds);
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply operators/src/source/gdal_source.rs]: 66f2ce9

**Review by michaelmattig — APPROVED** [review]

**Pull request merged**